# 제로베이스 데이터 사이언스 스쿨 EDA Level Test 8
## 문제 소개 
- 총 3단계 데이터 분석 상황 제시
<br></br>
    - 1단계: DataFrame 불러오기 및 전처리
    - 2단계: 원하는 정보 얻기 01 - 기초
    - 3단계: 원하는 정보 얻기 02 - 심화
<br></br>
    - 총점 100점
<br></br>
## 참고 사항
- Check Functions 셀을 반드시 실행해주세요!
- 여러분이 제출하신 답안을 채점하기 위한 코드입니다. 
- 각 문제를 푼 후 반드시, Check(예: check_01_01(df)) 함수를 실행해주시기를 바랍니다
<br></br>
- 파일 제목은 [DS]edatest_honggildong.ipynb 으로 제출해주시기 바랍니다.
    - 성부터 소문자로 작성해주세요.
- 제출은 제공드린 구글설문지를 통해 제출해주시면 감사하겠습니다.
---

# 1. Check Functions
- grading 채점 코드 안에는 각 문제 번호에 따른 check_00_00() 라는 함수가 있습니다. 

    ```
    from grading import check_01_01
    from grading import check_01_02
    from grading import check_01_03
    from grading import check_01_04
    from grading import check_01_05
    from grading import check_02_01
    from grading import check_02_02
    from grading import check_02_03
    from grading import check_02_04
    from grading import check_03_01
    from grading import check_03_02
    ```
    
- 답안을 작성하신 후에 반드시 채점 코드까지 결과값까지 포함하여 제출해주시기를 바랍니다.  

ex) 
check_01_01(df)

정답입니다! 10점 누적 되었습니다!

현재 누적 점수: 10 / 100

In [1]:
# 채점을 위한 코드입니다. 반드시 실행해주세요.
from grading import *

# 2. Datas

##### Target Data(csv): 서울시 지하철 호선별 역별 유/무임 승하차 인원 정보
- Source: [서울 열린데이터 광장](http://data.seoul.go.kr/dataList/OA-12251/S/1/datasetView.do)
- DownLoad: 상기 Source(서울 열린데이터 광장)에서 csv 파일 Download
    - 문제 출제 이후 Data 변경될 가능성이 있으므로 문제 풀이시 하기 File Path의 Data 사용
- File Path: ```'./datas/서울시 지하철 호선별 역별 유_무임 승하차 인원 정보.csv'```


##### 참고사항
- 문제에 hint가 있을 경우, 해당 hint를 이용하지 않으셔도 무방합니다.

# 3. 문제

## 1단계: DataFrame 불러오기 및 전처리
***

### 문제 1-1) DataFrame 불러오기 (5점)
- 위 2. Datas의 File Path를 참고하여 아래 조건에 맞게 해당 csv 파일을 Pandas DataFrame으로 불러오세요.
<br></br>
    - 조건1: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건2: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
    - hint: encoding 확인
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_target)를 check_01_01 함수에 입력하여 채점하세요.

In [113]:
import pandas
import numpy as np

In [114]:
df_target = pd.read_csv("./datas/서울시 지하철 호선별 역별 유_무임 승하차 인원 정보.csv",  sep=",", encoding='cp949')
df_target

사용월  호선명          지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자
0      202206  1호선           제기동   218782  263173   209109  281985  20220703
1      202206  1호선  청량리(서울시립대입구)   328288  246164   325365  249818  20220703
2      202206  1호선            종각   834173  131987   815704  121717  20220703
3      202206  1호선           서울역  1150754  194717  1123193  187155  20220703
4      202206  1호선          종로3가   450317  299617   404106  278151  20220703
...       ...  ...           ...      ...     ...      ...     ...       ...
60055  201501  9호선           신방화   169119  160054    38500   38259  20150206
60056  201501  9호선          양천향교   169461  165605    35179   35196  20150206
60057  201501  9호선            등촌   265941  271132    73910   74472  20150206
60058  201501  9호선            가양   469127  439698   107695  104082  20150206
60059  201501  9호선            당산   422698  384841    54179   49891  20150206

[60060 rows x 8 columns]

In [115]:
df_target.loc[620:626, :]

사용월  호선명  지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자
620  202205  1호선  종로3가   481950  311628   426618  287827  20220613
621  202205  1호선   제기동   231626  278255   222252  299053  20220613
622  202205  1호선  종로5가   408794  267599   416893  254913  20220613
623  202205  1호선   서울역  1190797  205902  1163681  197470  20220613
624  202205  1호선    종각   863844  143903   838956  132619  20220613
625  202205  1호선   서울역  1190797  205902  1163681  197470  20220613
626  202205  1호선  종로5가   408794  267599   416893  254913  20220613

In [116]:
check_01_01(df_target)

정답입니다! 5점 누적 되었습니다!
현재 누적 점수: 20 / 100


### 문제 1-2) DataFrame 전처리 01 - 중복제거 01 (5점)
- 중복되는 Data를 삭제하고자 합니다.
<br></br>
- 위 1-1의 DataFrame(df_target)을 아래 조건에 맞게 변경하세요.
<br></br>
    - 조건1: 각 Row(행) 중 ['사용월', '호선명', '지하철역', '유임승차인원', '무임승차인원', '유임하차인원', '무임하차인원'] 의 내용이 모두 똑같은(중복되는) Row(행)이 있다면 가장 앞의 Row(Index가 가장 낮은 Row(행))는 남기고 나머지 중복되는 Row(행)를 삭제하세요.
<br></br>
        - 예시1:  각 Row(행) 중 ['사용월', '호선명', '지하철역', '유임승차인원', '무임승차인원', '유임하차인원', '무임하차인원'] 의 내용이 모두 똑같은(중복되는) Row(행)
<br></br>
        <img src="./[문제보기용 이미지]images/08-1-2-001.png" style="margin:auto" width=600/>
<br></br>
        - 예시2: 가장 앞의 Row(Index가 가장 낮은 Row(행), 621)는 남기고 나머지 중복되는 Row(627)를 삭제
<br></br>
        <img src="./[문제보기용 이미지]images/08-1-2-002.png" style="margin:auto" width=600/>
<br></br>
    - 조건2: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건3: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_target)를 check_01_02 함수에 입력하여 채점하세요.

In [117]:
df_target = df_target.drop_duplicates(subset=['사용월', '호선명', '지하철역', '유임승차인원', \
                                              '무임승차인원', '유임하차인원', '무임하차인원'])
df_target

사용월  호선명          지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자
0      202206  1호선           제기동   218782  263173   209109  281985  20220703
1      202206  1호선  청량리(서울시립대입구)   328288  246164   325365  249818  20220703
2      202206  1호선            종각   834173  131987   815704  121717  20220703
3      202206  1호선           서울역  1150754  194717  1123193  187155  20220703
4      202206  1호선          종로3가   450317  299617   404106  278151  20220703
...       ...  ...           ...      ...     ...      ...     ...       ...
60055  201501  9호선           신방화   169119  160054    38500   38259  20150206
60056  201501  9호선          양천향교   169461  165605    35179   35196  20150206
60057  201501  9호선            등촌   265941  271132    73910   74472  20150206
60058  201501  9호선            가양   469127  439698   107695  104082  20150206
60059  201501  9호선            당산   422698  384841    54179   49891  20150206

[51162 rows x 8 columns]

In [118]:
check_01_02(df_target)

정답입니다! 5점 누적 되었습니다!
현재 누적 점수: 20 / 100


### 문제 1-3) DataFrame 전처리 02 - 중복제거 02 (10점)
- 위 1-2에서 삭제되지 않은 중복되는 Data를 삭제하고자 합니다.
<br></br>
- 위 1-2의 DataFrame(df_target)을 아래 조건에 맞게 변경하세요.
<br></br>
    - 조건1: 각 Row(행) 중 ['사용월', '호선명', '지하철역'] 의 내용이 모두 똑같은(중복되는) Row(행)이 있다면 가장 앞의 Row(Index가 가장 낮은 Row(행))에 ['유임승차인원', '무임승차인원', '유임하차인원', '무임하차인원'] 각각 합한 값을 입력하고, 나머지 ['사용월', '호선명', '지하철역'] 의 내용이 중복되는 Row(행)를 삭제하세요.
<br></br>
        - 예시1: ['사용월', '호선명', '지하철역'] 의 내용이 모두 똑같은(중복되는) Row(행)
<br></br>
        <img src="./[문제보기용 이미지]images/08-1-3-001.png" style="margin:auto" width=600/>
<br></br>
        - 예시2: 가장 앞의 Row(Index가 가장 낮은 Row(행), 15585)에 ['유임승차인원', '무임승차인원', '유임하차인원', '무임하차인원'] 각각 합한 값을 입력
<br></br>
        <img src="./[문제보기용 이미지]images/08-1-3-002.png" style="margin:auto" width=600/>
<br></br>
        - 예시3: 나머지 ['사용월', '호선명', '지하철역'] 의 내용이 중복되는 Row(행, 15595)를 삭제
<br></br>
        <img src="./[문제보기용 이미지]images/08-1-3-003.png" style="margin:auto" width=600/>
<br></br>
    - 조건2: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건3: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_target)를 check_01_03 함수에 입력하여 채점하세요.

In [119]:
df_target[(df_target["사용월"] == 202006) & (df_target["호선명"] == '경의선') & (df_target["지하철역"] == '디지털미디어시티')]

사용월  호선명      지하철역  유임승차인원  무임승차인원  유임하차인원  무임하차인원      작업일자
15585  202006  경의선  디지털미디어시티       1       0       0       0  20200703
15595  202006  경의선  디지털미디어시티   52699    8612   49913    8920  20200703

In [120]:
df_target[(df_target["사용월"] == 202006) & (df_target["호선명"] == '경의선') & (df_target["지하철역"] == '디지털미디어시티')].\
groupby(['사용월', '호선명', '지하철역']).sum()

유임승차인원  무임승차인원  유임하차인원  무임하차인원      작업일자
사용월    호선명 지하철역                                              
202006 경의선 디지털미디어시티   52700    8612   49913    8920  40401406

In [121]:
df_target[df_target.duplicated(subset=['사용월', '호선명', '지하철역'], keep='first')==False].\
sort_values(by=['사용월', '호선명', '지하철역']).reset_index()

index     사용월  호선명 지하철역   유임승차인원   무임승차인원  유임하차인원  무임하차인원      작업일자
0      59761  201501  1호선  동대문   336981   390926  132944  137814  20150206
1      59760  201501  1호선  동묘앞   155556   171171  124838  124116  20150206
2      59767  201501  1호선  서울역  1890411  1667163  238403  220008  20150206
3      59765  201501  1호선   시청   639441   636014   87392   85689  20150206
4      59766  201501  1호선  신설동   374981   363032  124078  118929  20150206
...      ...     ...  ...  ...      ...      ...     ...     ...       ...
51087    606  202206  중앙선   원덕     5854     6959    5749    6739  20220703
51088    605  202206  중앙선   중랑   115112    47960  106287   46789  20220703
51089    610  202206  중앙선   지평      952     1216    1282    1173  20220703
51090    596  202206  중앙선   팔당    23038     9401   24414    9398  20220703
51091    612  202206  중앙선   회기   597740   121604  572831  119310  20220703

[51092 rows x 9 columns]

In [122]:
index_num = df_target[df_target.duplicated(subset=['사용월', '호선명', '지하철역'], keep='first')==False].\
sort_values(by=['사용월', '호선명', '지하철역']).reset_index()['index']

date_num = df_target[df_target.duplicated(subset=['사용월', '호선명', '지하철역'], keep='first')==False].\
sort_values(by=['사용월', '호선명', '지하철역']).reset_index()['작업일자']

In [123]:
df_target = df_target.groupby(['사용월', '호선명', '지하철역'], as_index=False).sum()
df_target

사용월  호선명 지하철역   유임승차인원   무임승차인원  유임하차인원  무임하차인원      작업일자
0      201501  1호선  동대문   336981   390926  132944  137814  20150206
1      201501  1호선  동묘앞   155556   171171  124838  124116  20150206
2      201501  1호선  서울역  1890411  1667163  238403  220008  20150206
3      201501  1호선   시청   639441   636014   87392   85689  20150206
4      201501  1호선  신설동   374981   363032  124078  118929  20150206
...       ...  ...  ...      ...      ...     ...     ...       ...
51087  202206  중앙선   원덕     5854     6959    5749    6739  20220703
51088  202206  중앙선   중랑   115112    47960  106287   46789  20220703
51089  202206  중앙선   지평      952     1216    1282    1173  20220703
51090  202206  중앙선   팔당    23038     9401   24414    9398  20220703
51091  202206  중앙선   회기   597740   121604  572831  119310  20220703

[51092 rows x 8 columns]

In [124]:
df_target = df_target.sort_values(by=['사용월', '호선명', '지하철역'])
df_target["작업일자"] = date_num
df_target['index'] = index_num
df_target

사용월  호선명 지하철역   유임승차인원   무임승차인원  유임하차인원  무임하차인원      작업일자  index
0      201501  1호선  동대문   336981   390926  132944  137814  20150206  59761
1      201501  1호선  동묘앞   155556   171171  124838  124116  20150206  59760
2      201501  1호선  서울역  1890411  1667163  238403  220008  20150206  59767
3      201501  1호선   시청   639441   636014   87392   85689  20150206  59765
4      201501  1호선  신설동   374981   363032  124078  118929  20150206  59766
...       ...  ...  ...      ...      ...     ...     ...       ...    ...
51087  202206  중앙선   원덕     5854     6959    5749    6739  20220703    606
51088  202206  중앙선   중랑   115112    47960  106287   46789  20220703    605
51089  202206  중앙선   지평      952     1216    1282    1173  20220703    610
51090  202206  중앙선   팔당    23038     9401   24414    9398  20220703    596
51091  202206  중앙선   회기   597740   121604  572831  119310  20220703    612

[51092 rows x 9 columns]

In [125]:
df_target = df_target.sort_values(by="index")
df_target = df_target.set_index("index")
df_target

사용월  호선명          지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자
index                                                                       
0      202206  1호선           제기동   218782  263173   209109  281985  20220703
1      202206  1호선  청량리(서울시립대입구)   328288  246164   325365  249818  20220703
2      202206  1호선            종각   834173  131987   815704  121717  20220703
3      202206  1호선           서울역  1150754  194717  1123193  187155  20220703
4      202206  1호선          종로3가   450317  299617   404106  278151  20220703
...       ...  ...           ...      ...     ...      ...     ...       ...
60055  201501  9호선           신방화   169119  160054    38500   38259  20150206
60056  201501  9호선          양천향교   169461  165605    35179   35196  20150206
60057  201501  9호선            등촌   265941  271132    73910   74472  20150206
60058  201501  9호선            가양   469127  439698   107695  104082  20150206
60059  201501  9호선            당산   422698  384841    54179   49891  20150206

[51092 rows x 8 columns]

In [126]:
check_01_03(df_target)

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 20 / 100


### 문제 1-4) DataFrame 전처리 03 - Data 수정 01 (5점)
- 위 1-3의 DataFrame(df_target)을 아래 조건에 맞게 변경하세요.
<br></br>
    - 조건1: '호선명' Column(열)의 Data가 '9호선2~3단계' 또는 '9호선2단계'라면 이를 '9호선'으로 변경하세요.
<br></br>
    - 조건2: '지하철역' Column(열)의 Data에 괄호('(', ')')가 있다면, 괄호 안의 내용과 괄호 자체를 모두 지워주세요.
        - 예시: '청량리(서울시립대입구)' -> '청량리'
<br></br>
    - 조건3: '지하철역' Column(열)의 Data가 '신천'이라면 '잠실새내'로 변경하세요.
<br></br>
    - 조건4: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건5: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_target)를 check_01_04 함수에 입력하여 채점하세요.

In [127]:
df_target['호선명'].unique()

array(['1호선', '2호선', '3호선', '4호선', '5호선', '6호선', '7호선', '8호선', '9호선',
       '9호선2~3단계', '경강선', '경부선', '경원선', '경의선', '경인선', '경춘선', '공항철도 1호선',
       '과천선', '분당선', '수인선', '신림선', '안산선', '우이신설선', '일산선', '장항선', '중앙선',
       '9호선2단계'], dtype=object)

In [128]:
df_target['호선명'] = df_target['호선명'].str.replace("9호선2~3단계", "9호선")
df_target['호선명'] = df_target['호선명'].str.replace("9호선2단계", "9호선")
df_target['호선명'].unique()

array(['1호선', '2호선', '3호선', '4호선', '5호선', '6호선', '7호선', '8호선', '9호선',
       '경강선', '경부선', '경원선', '경의선', '경인선', '경춘선', '공항철도 1호선', '과천선', '분당선',
       '수인선', '신림선', '안산선', '우이신설선', '일산선', '장항선', '중앙선'], dtype=object)

In [138]:
df_target['지하철역']=df_target['지하철역'].str.replace(r"\(.*\)","")

C:\Users\xswer\AppData\Local\Temp\ipykernel_21632\2460577833.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df_target['지하철역']=df_target['지하철역'].str.replace(r"\(.*\)","")


In [139]:
df_target['지하철역'] = df_target['지하철역'].str.replace("신천", '잠실새내')

In [140]:
check_01_04(df_target)

정답입니다! 5점 누적 되었습니다!
현재 누적 점수: 25 / 100


### 문제 1-5) DataFrame 전처리 03 - Data 수정 02 (10점)
- 아래 Cell의 incheon_station_list에 있는 역은 모두 7호선의 역입니다. 해당 역들은 서울특별시가 아닌 인천광역시(인천교통공사)가 담당하여 Data가 부정확한 것으로 추정됩니다. 이에 기존 7호선과 인천광역시 담당 7호선을 구분하려고 합니다.
<br></br>
- 위 1-4의 DataFrame(df_target)을 아래 조건에 맞게 변경하세요.
<br></br>
    - 조건1: incheon_station_list에 있는 역들의 '호선명'을 '7호선'에서 '7호선 인천'으로 변경해주세요.
<br></br>
        - 예시1: '7호선' 변경 전
<br></br>
        <img src="./[문제보기용 이미지]images/08-1-5-001.png" style="margin:auto" width=600/>
<br></br>
        - 예시2: '7호선 인천'으로 변경 후
<br></br>
        <img src="./[문제보기용 이미지]images/08-1-5-002.png" style="margin:auto" width=600/>
<br></br>
    - 조건2: '지하철역' Column(열)의 Data에 괄호('(', ')')가 있다면, 괄호 안의 내용과 괄호 자체를 모두 지워주세요.
        - 예시: '청량리(서울시립대입구)' -> '청량리'
<br></br>
    - 조건1: '지하철역' Column(열)의 Data가 '신천'이라면 '잠실새내'로 변경하세요.
<br></br>
    - 조건2: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건3: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_target)를 check_01_05 함수에 입력하여 채점하세요.

In [144]:
incheon_station_list = ['까치울', '부천종합운동장', '춘의', '신중동', '부천시청', '상동', \
                        '삼산체육관', '굴포천', '부평구청', '산곡', '석남']

In [149]:
index_list = []
for name in incheon_station_list:
    index_num = df_target[(df_target["호선명"]=="7호선") & (df_target['지하철역']==name)].index
    index_list.append(index_num)

In [160]:
def flatten(list):
    result = []
    for item in list:
        result.extend(item)
    return result

In [162]:
index_list = flatten(index_list)
index_list

[247,
 1102,
 2081,
 2698,
 3318,
 3926,
 4506,
 5122,
 5756,
 6365,
 6972,
 7579,
 8186,
 8794,
 9401,
 10009,
 10610,
 11215,
 11817,
 12422,
 13025,
 13630,
 14223,
 14819,
 15403,
 15994,
 16608,
 17220,
 17820,
 18380,
 18978,
 19614,
 20170,
 20794,
 21409,
 22394,
 25655,
 29717,
 30307,
 30935,
 31530,
 32150,
 32710,
 33316,
 33914,
 34487,
 35078,
 35669,
 36255,
 36846,
 37434,
 38238,
 39198,
 39790,
 40374,
 40997,
 41548,
 42126,
 42713,
 43282,
 43862,
 44421,
 44990,
 45558,
 46129,
 46714,
 47266,
 47869,
 48647,
 49775,
 50700,
 51274,
 51829,
 52366,
 52912,
 53503,
 54045,
 54594,
 55140,
 55699,
 56263,
 56806,
 57373,
 57909,
 58464,
 58764,
 59066,
 59386,
 59683,
 60011,
 2088,
 2701,
 3305,
 3913,
 4531,
 5135,
 5749,
 6357,
 6964,
 7571,
 8178,
 8786,
 9394,
 10002,
 10582,
 11187,
 11791,
 12394,
 12997,
 13602,
 14204,
 14812,
 15387,
 16014,
 16631,
 17229,
 17802,
 18394,
 18977,
 19579,
 20174,
 20779,
 21385,
 22411,
 25658,
 29736,
 30357,
 30947,
 3151

In [163]:
df_target.loc[index_list, "호선명"] = '7호선 인천'

In [164]:
df_target[(df_target["호선명"]=="7호선 인천") & (df_target["지하철역"]=="까치울")].index

Int64Index([  247,  1102,  2081,  2698,  3318,  3926,  4506,  5122,  5756,
             6365,  6972,  7579,  8186,  8794,  9401, 10009, 10610, 11215,
            11817, 12422, 13025, 13630, 14223, 14819, 15403, 15994, 16608,
            17220, 17820, 18380, 18978, 19614, 20170, 20794, 21409, 22394,
            25655, 29717, 30307, 30935, 31530, 32150, 32710, 33316, 33914,
            34487, 35078, 35669, 36255, 36846, 37434, 38238, 39198, 39790,
            40374, 40997, 41548, 42126, 42713, 43282, 43862, 44421, 44990,
            45558, 46129, 46714, 47266, 47869, 48647, 49775, 50700, 51274,
            51829, 52366, 52912, 53503, 54045, 54594, 55140, 55699, 56263,
            56806, 57373, 57909, 58464, 58764, 59066, 59386, 59683, 60011],
           dtype='int64', name='index')

In [165]:
check_01_05(df_target)

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 35 / 100


## 2단계: 원하는 정보 얻기 01 - 기초
***

### 문제 2-1) 기초 정보 얻기 01 (5점)
- 위 1단계의 DataFrame(df_target)을 이용하여 아래 예시 및 조건에 맞는 정보를 구하세요.
<br></br>
    - 예시: df_result.iloc[:2]
<br></br>
        <img src="./[문제보기용 이미지]images/08-2-1-001.png" style="margin:auto" width=400/>
<br></br>
<br></br>
    - 조건1: 2022년 6월 기준 각 지하철 역의 [유임승차인원, 무임승차인원, 유임하차인원, 무임하차인원] 을 구하세요.
<br></br>
    - 조건2: 환승역(한 지하철역에 여러개의 호선이 있는 역)이 있다면 합산해 주세요.
<br></br>
        - 예시: 종합운동장역의 경우 2, 9호선 유/무임 승/하차인원을 합쳐서 보여주세요.
<br></br>
        <img src="./[문제보기용 이미지]images/08-2-1-002.png" style="margin:auto" width=600/>
<br></br>
        <img src="./[문제보기용 이미지]images/08-2-1-003.png" style="margin:auto" width=400/>
<br></br>
    - 조건3: Index와 Column은 아래와 같이 지정해 주세요.
        - Index: '지하철역'
        - Column: ['유임승차인원', '무임승차인원', '유임하차인원', '무임하차인원']
<br></br>
    - 조건4: Index 기준 오름차순으로 정렬해주세요.
<br></br>
    - 조건5: 'df_result' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_result)를 check_02_01 함수에 입력하여 채점하세요.

In [175]:
df_target

사용월  호선명  지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자
index                                                               
0      202206  1호선   제기동   218782  263173   209109  281985  20220703
1      202206  1호선   청량리   328288  246164   325365  249818  20220703
2      202206  1호선    종각   834173  131987   815704  121717  20220703
3      202206  1호선   서울역  1150754  194717  1123193  187155  20220703
4      202206  1호선  종로3가   450317  299617   404106  278151  20220703
...       ...  ...   ...      ...     ...      ...     ...       ...
60055  201501  9호선   신방화   169119  160054    38500   38259  20150206
60056  201501  9호선  양천향교   169461  165605    35179   35196  20150206
60057  201501  9호선    등촌   265941  271132    73910   74472  20150206
60058  201501  9호선    가양   469127  439698   107695  104082  20150206
60059  201501  9호선    당산   422698  384841    54179   49891  20150206

[51092 rows x 8 columns]

In [176]:
df_target[(df_target["사용월"] == 202206)]

사용월  호선명  지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자
index                                                               
0      202206  1호선   제기동   218782  263173   209109  281985  20220703
1      202206  1호선   청량리   328288  246164   325365  249818  20220703
2      202206  1호선    종각   834173  131987   815704  121717  20220703
3      202206  1호선   서울역  1150754  194717  1123193  187155  20220703
4      202206  1호선  종로3가   450317  299617   404106  278151  20220703
...       ...  ...   ...      ...     ...      ...     ...       ...
610    202206  중앙선    지평      952    1216     1282    1173  20220703
611    202206  중앙선    도심    51755   18705    38500   17600  20220703
612    202206  중앙선    회기   597740  121604   572831  119310  20220703
613    202206  중앙선    양정    13353    8004    10920    7087  20220703
614    202206  중앙선    신원     3713    3699     3154    3140  20220703

[615 rows x 8 columns]

In [178]:
df_2_1= df_target[(df_target["사용월"] == 202206)].groupby("지하철역").sum().\
drop(columns=['작업일자','사용월']).sort_index()

C:\Users\xswer\AppData\Local\Temp\ipykernel_21632\1873875035.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df_2_1= df_target[(df_target["사용월"] == 202206)].groupby("지하철역").sum().\


In [180]:
df_2_1

유임승차인원  무임승차인원   유임하차인원  무임하차인원
지하철역                                      
4.19민주묘지    56275   34889    54138   30800
가능         139026   55693   131155   54568
가락시장       322711  129590   342636  128815
가산디지털단지   1431139  117639  1502220  117128
가양         456785  123819   442530  120665
...           ...     ...      ...     ...
회기         597740  121604   572831  119310
회룡         222825  142728   219069  141480
회현         490517  191008   523086  191242
효창공원앞      231459   50062   222409   48748
흑석         186994   53158   191190   53327

[524 rows x 4 columns]

In [179]:
check_02_01(df_2_1)

정답입니다! 5점 누적 되었습니다!
현재 누적 점수: 40 / 100


### 문제 2-2) 기초 정보 얻기 02 (10점)
- 위 1단계의 DataFrame(df_target)을 이용하여 아래 예시 및 조건에 맞는 정보를 구하세요.
<br></br>
    - 예시: df_result.iloc[:4]
<br></br>
        <img src="./[문제보기용 이미지]images/08-2-2-001.png" style="margin:auto" width=100/>
<br></br>
    - 조건1: 2022년 6월 기준 각 지하철 역의 호선의 수를 구하고, 호선의 수가 3개 이상인 역만 보여주세요.
<br></br>
    - 조건2: '호선수' Column(열)을 만들어 해당 역의 호선 수를 입력해주세요.
<br></br>
    - 조건3: Index와 Column은 아래와 같이 지정해 주세요.
        - Index: '지하철역'
        - Column: ['호선수']
<br></br>
    - 조건4: '호선수' Column(열)을 기준으로 내림차순 정렬하고, '호선수'가 같다면 Index('지하철역')를 기준으로 내림차순 정렬해주세요.
<br></br>
    - 조건5: 'df_result' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_result)를 check_02_02 함수에 입력하여 채점하세요.

In [213]:
len(df_target[(df_target['사용월']==202206) & (df_target['지하철역']=='종로3가')]['호선명'].unique())

3

In [214]:
station_list = df_target["지하철역"].unique()
len(station_list)

533

In [215]:
station_name = []
station_count = []

for name in station_list:
    station_name.append(name)
    station_count.append(len(df_target[(df_target['사용월']==202206) & (df_target['지하철역']==name)]['호선명'].unique()))

In [216]:
len(station_name), len(station_count)

(533, 533)

In [217]:
station_count

[1,
 2,
 1,
 5,
 3,
 1,
 2,
 2,
 2,
 3,
 2,
 3,
 1,
 2,
 2,
 1,
 1,
 2,
 2,
 1,
 1,
 1,
 2,
 2,
 1,
 1,
 1,
 1,
 1,
 3,
 1,
 2,
 2,
 1,
 2,
 1,
 2,
 2,
 3,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 2,
 1,
 2,
 1,
 2,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 2,
 1,
 2,
 1,
 1,
 2,
 2,
 1,
 2,
 1,
 1,
 1,
 1,
 2,
 3,
 1,
 1,
 2,
 1,
 2,
 1,
 2,
 1,
 1,
 2,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 2,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 4,
 2,
 1,
 1,
 1,
 1,
 1,
 3,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 2,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 3,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 1,
 1,
 1,
 2,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 2,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,


In [218]:
df_2_2 = pd.DataFrame(index=station_name)
df_2_2["호선수"] = station_count
df_2_2

호선수
제기동       1
청량리       2
종각        1
서울역       5
종로3가      3
...     ...
춘의        0
지제        0
인천국제공항    0
동두천 중앙    0
쌍용동       0

[533 rows x 1 columns]

In [219]:
df_2_2 = df_2_2[df_2_2['호선수'] >= 3].reset_index().sort_values(by=['호선수', 'index'], ascending=False).\
set_index("index", drop=True)
df_2_2

호선수
index         
서울역          5
공덕           4
홍대입구         3
종로3가         3
왕십리          3
신설동          3
디지털미디어시티     3
동대문역사문화공원    3
김포공항         3
고속터미널        3

In [220]:
check_02_02(df_2_2)

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 50 / 100


### 문제 2-3) 기초 정보 얻기 03 (10점)
- 위 1단계의 DataFrame(df_target)을 이용하고, 위 2-2의 DataFrame(df_result)를 응용하여 아래 예시 및 조건에 맞는 정보를 구하세요.
<br></br>
    - 예시: df_result.iloc[:4]
<br></br>
        <img src="./[문제보기용 이미지]images/08-2-3-001.png" style="margin:auto" width=400/>
<br></br>
    - 조건1: 위 2-2를 참고하여 아래와 같이 '호선수'를 구하세요.
        - 2022년 6월 기준 각 지하철 역의 호선의 수를 구하고, 호선의 수가 3개 이상인 역만 보여주세요.
        - '호선수' Column(열)을 만들어 해당 역의 호선 수를 입력해주세요.
<br></br>
    - 조건2: 2022년 6월 기준 '환승호선' Column(열)을 만들어 아래와 같이 해당 역의 환승 가능 호선을 나열해주세요.
        - 환승 호선들은 ', '(쉼표+띄어쓰기)로 구분되는 string입니다.
        - 환승 호선들은 오름차순으로 정렬 후 나열해주세요.
        - 위 예시를 참고해주세요.
<br></br>
    - 조건3: Index와 Column은 아래와 같이 지정해 주세요.
        - Index: '지하철역'
        - Column: ['호선수', '환승호선']
<br></br>
    - 조건4: '호선수' Column(열)을 기준으로 내림차순 정렬하고, '호선수'가 같다면 '환승호선' Column(열)을 기준으로 오름차순 정렬해주세요.
<br></br>
    - 조건5: 'df_result' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_result)를 check_02_03 함수에 입력하여 채점하세요.

In [238]:
', '.join(df_target[(df_target['사용월']==202206) & (df_target['지하철역']=='공덕')]['호선명'].unique())

'5호선, 6호선, 경의선, 공항철도 1호선'

In [223]:
station_list = df_target["지하철역"].unique()
len(station_list)

533

In [239]:
station_name = []
station_count = []
station_line = []

for name in station_list:
    station_name.append(name)
    station_count.append(len(df_target[(df_target['사용월']==202206) & (df_target['지하철역']==name)]['호선명'].unique()))
    station_line.append(', '.join(df_target[(df_target['사용월']==202206) & (df_target['지하철역']==name)]['호선명'].unique()))

In [240]:
len(station_name), len(station_count), len(station_line)

(533, 533, 533)

In [241]:
df_2_3 = pd.DataFrame(index=station_name)
df_2_3["호선수"] = station_count
df_2_3["환승호선"] = station_line
df_2_3

호선수                          환승호선
제기동       1                           1호선
청량리       2                      1호선, 경원선
종각        1                           1호선
서울역       5  1호선, 4호선, 경부선, 경의선, 공항철도 1호선
종로3가      3                 1호선, 3호선, 5호선
...     ...                           ...
춘의        0                              
지제        0                              
인천국제공항    0                              
동두천 중앙    0                              
쌍용동       0                              

[533 rows x 2 columns]

In [244]:
df_2_3 = df_2_3.loc[df_2_3["호선수"]>=3].sort_values(by=['호선수', '환승호선'], ascending=[False, True])
df_2_3

호선수                          환승호선
서울역          5  1호선, 4호선, 경부선, 경의선, 공항철도 1호선
공덕           4       5호선, 6호선, 경의선, 공항철도 1호선
신설동          3               1호선, 2호선, 우이신설선
종로3가         3                 1호선, 3호선, 5호선
동대문역사문화공원    3                 2호선, 4호선, 5호선
왕십리          3                 2호선, 5호선, 경원선
홍대입구         3            2호선, 경의선, 공항철도 1호선
고속터미널        3                 3호선, 7호선, 9호선
김포공항         3            5호선, 9호선, 공항철도 1호선
디지털미디어시티     3            6호선, 경의선, 공항철도 1호선

In [245]:
check_02_03(df_2_3)

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 60 / 100


### 문제 2-4) 기초 정보 얻기 04 (10점)
- 위 1단계의 DataFrame(df_target)을 이용하고, 위 2-2의 DataFrame(df_result)를 응용하여 아래 예시 및 조건에 맞는 정보를 구하세요.
<br></br>
    - 예시: df_result.iloc[:4]
<br></br>
        <img src="./[문제보기용 이미지]images/08-2-4-001.png" style="margin:auto" width=300/>
<br></br>
    - 조건1: 위 2-2를 참고하여 아래와 같이 '호선수'를 구하세요.
        - 2022년 6월 기준 각 지하철 역의 호선의 수를 구하고, 호선의 수가 2개 이상인 역만 보여주세요.
        - '호선수' Column(열)을 만들어 해당 역의 호선 수를 입력해주세요.
<br></br>
    - 조건2: 2022년 6월 기준 ['유임승하차율', '무임승하차율', '승하차율'] Column(열)을 만들어 아래와 같이 각각의 승하차율을 구하세요.
        - 유임승하차율 = 유임승차인원 / 유임하차인원 * 100
        - 무임승하차율 = 무임승차인원 / 무임하차인원 * 100
        - 승하차율 = (유임승차인원 + 무임승차인원) / (유임하차인원 + 무임하차인원) * 100
        - 각각의 승하차율은 소숫점 셋째자리에서 반올림하여 둘째자리까지 표기합니다.
        - '무임승하차율'이 '승하차율'보다 크면서, '무임승하차율'이 102 이상인 역만 보여주세요.
        - 위 예시를 참고해주세요.
<br></br>
    - 조건3: Index와 Column은 아래와 같이 지정해 주세요.
        - Index: '지하철역'
        - Column: ['호선수', '유임승하차율', '무임승하차율', '승하차율']
<br></br>
    - 조건4: '호선수' Column(열)을 기준으로 내림차순 정렬하고, '호선수'가 같다면 '무임승하차율' Column(열)을 기준으로 내림차순 정렬해주세요.
<br></br>
    - 조건5: 'df_result' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_result)를 check_02_04 함수에 입력하여 채점하세요.

In [281]:
df_2_2[df_2_2["호선수"]>=3]

호선수
index         
서울역          5
공덕           4
홍대입구         3
종로3가         3
왕십리          3
신설동          3
디지털미디어시티     3
동대문역사문화공원    3
김포공항         3
고속터미널        3

In [269]:
station_list = df_target["지하철역"].unique()
len(station_list)

533

In [270]:
df_target[(df_target['사용월']==202206) & (df_target['지하철역']=='서울역')]['유임승차인원'].sum()

2009351

In [271]:
station_name = []
station_count = []
station_1 = []
station_2 = []
station_3 = []
station_4 = []


for name in station_list:
    station_name.append(name)
    station_count.append(len(df_target[(df_target['사용월']==202206) & (df_target['지하철역']==name)]['호선명'].unique()))
    station_1.append(df_target[(df_target['사용월']==202206) & (df_target['지하철역']==name)]['유임승차인원'].sum())
    station_2.append(df_target[(df_target['사용월']==202206) & (df_target['지하철역']==name)]['무임승차인원'].sum())
    station_3.append(df_target[(df_target['사용월']==202206) & (df_target['지하철역']==name)]['유임하차인원'].sum())
    station_4.append(df_target[(df_target['사용월']==202206) & (df_target['지하철역']==name)]['무임하차인원'].sum())

In [272]:
df_2_4 = pd.DataFrame(index=station_name)
df_2_4["호선수"] = station_count
df_2_4["유임승차인원"] = station_1
df_2_4["무임승차인원"] = station_2
df_2_4["유임하차인원"] = station_3
df_2_4["무임하차인원"] = station_4
df_2_4

호선수   유임승차인원  무임승차인원   유임하차인원  무임하차인원
제기동       1   218782  263173   209109  281985
청량리       2   677433  390621   677748  399692
종각        1   834173  131987   815704  121717
서울역       5  2009351  296197  2036065  287745
종로3가      3   960097  449445   954987  437758
...     ...      ...     ...      ...     ...
춘의        0        0       0        0       0
지제        0        0       0        0       0
인천국제공항    0        0       0        0       0
동두천 중앙    0        0       0        0       0
쌍용동       0        0       0        0       0

[533 rows x 5 columns]

In [273]:
df_2_4 = df_2_4[df_2_4['호선수']>=2]
df_2_4

호선수   유임승차인원  무임승차인원   유임하차인원  무임하차인원
청량리     2   677433  390621   677748  399692
서울역     5  2009351  296197  2036065  287745
종로3가    3   960097  449445   954987  437758
동대문     2   603756  269974   607364  265534
시청      2  1082621  132013  1070141  124357
...   ...      ...     ...      ...     ...
수원      2  1181915  176602  1208695  175840
대방      2   294831   71372   304573   70700
검암      2   171189   16827   169675   16966
계양      2   274930   30703   253928   28727
인천      2   106366   56777   102341   56583

[78 rows x 5 columns]

In [274]:
df_2_4['유임승하차율'] = round((df_2_4['유임승차인원'] / df_2_4['유임하차인원']) * 100,2)
df_2_4['무임승하차율'] =round((df_2_4['무임승차인원'] / df_2_4['무임하차인원']) * 100,2)
df_2_4['승하차율'] = round(((df_2_4['유임승차인원'] + df_2_4['무임승차인원']) / (df_2_4['유임하차인원'] + df_2_4['무임하차인원']))\
                       * 100, 2)
df_2_4.drop(columns=["유임승차인원","무임승차인원","유임하차인원","무임하차인원",], inplace=True)

C:\Users\xswer\AppData\Local\Temp\ipykernel_21632\2625771855.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_2_4['유임승하차율'] = round((df_2_4['유임승차인원'] / df_2_4['유임하차인원']) * 100,2)
C:\Users\xswer\AppData\Local\Temp\ipykernel_21632\2625771855.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_2_4['무임승하차율'] =round((df_2_4['무임승차인원'] / df_2_4['무임하차인원']) * 100,2)
C:\Users\xswer\AppData\Local\Temp\ipykernel_21632\2625771855.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice fr

In [285]:
df_2_4 = df_2_4[(df_2_4["무임승하차율"]>df_2_4['승하차율']) & (df_2_4['무임승하차율']>=102)]
df_2_4

호선수  유임승하차율  무임승하차율    승하차율
서울역       5   98.69  102.94   99.21
공덕        4   97.96  104.65   98.79
홍대입구      3   92.93  103.17   93.43
종로3가      3  100.54  102.67  101.21
고속터미널     3   99.67  102.58  100.07
왕십리       3  101.06  102.29  101.22
여의도       2   97.69  106.58   98.23
시청        2  101.17  106.16  101.69
청구        2  102.47  105.86  103.14
삼각지       2  101.20  105.85  101.74
올림픽공원     2  104.39  105.25  104.55
도곡        2   99.71  105.15  100.79
이촌        2  100.87  105.01  101.73
약수        2  101.73  104.95  102.40
잠실        2   98.08  104.49   98.72
선릉        2   98.43  104.26   99.18
보라매       2  102.35  103.77  102.60
교대        2  100.96  103.64  101.41
태릉입구      2   99.43  103.43  100.12
신도림       2   98.65  102.74   99.24
성신여대입구    2  101.68  102.65  101.83
을지로3가     2   98.09  102.58   98.67
합정        2   94.15  102.36   94.77

In [286]:
df_2_4 = df_2_4.sort_values(by=["호선수", "무임승하차율"], ascending=False)
df_2_4

호선수  유임승하차율  무임승하차율    승하차율
서울역       5   98.69  102.94   99.21
공덕        4   97.96  104.65   98.79
홍대입구      3   92.93  103.17   93.43
종로3가      3  100.54  102.67  101.21
고속터미널     3   99.67  102.58  100.07
왕십리       3  101.06  102.29  101.22
여의도       2   97.69  106.58   98.23
시청        2  101.17  106.16  101.69
청구        2  102.47  105.86  103.14
삼각지       2  101.20  105.85  101.74
올림픽공원     2  104.39  105.25  104.55
도곡        2   99.71  105.15  100.79
이촌        2  100.87  105.01  101.73
약수        2  101.73  104.95  102.40
잠실        2   98.08  104.49   98.72
선릉        2   98.43  104.26   99.18
보라매       2  102.35  103.77  102.60
교대        2  100.96  103.64  101.41
태릉입구      2   99.43  103.43  100.12
신도림       2   98.65  102.74   99.24
성신여대입구    2  101.68  102.65  101.83
을지로3가     2   98.09  102.58   98.67
합정        2   94.15  102.36   94.77

In [287]:
check_02_04(df_2_4)

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 70 / 100


## 3단계: 원하는 정보 얻기 02 - 심화
***

### 문제 3-1) 심화 정보 얻기 01 (15점)
- 2022년 1월 대비 2022년 6월에 유동인구가 얼마나 증가/감소하였는지 확인해보고자 합니다.
<br></br>
- 위 1단계의 DataFrame(df_target)을 이용하여 아래 예시 및 조건에 맞는 정보를 구하세요.
<br></br>
    - 예시: df_result.iloc[:2]
<br></br>
        <img src="./[문제보기용 이미지]images/08-3-1-001.png" style="margin:auto" width=150/>
<br></br>
    - 조건1: '유동인구증감률' Column(열)을 만들어 2022년 1월 대비 2022년 6월의 유동인구 증가율을 입력하세요.
        - 유동인구 = '유임승차인원' + '무임승차인원' + '유임하차인원' + '무임하차인원'
        - 유동인구 증감률 = (2022년 6월의 유동인구 / 2022년 1월의 유동인구 - 1) * 100
        - 유동인구 증감률은 소숫점 셋째자리에서 반올림하여 둘째자리까지 표기합니다.
        - 유동인구가 감소한(유동인구 증감률이 0보다 작은) 역만 보여주세요.
<br></br>
    - 조건2: 2022년 1월 이후에 역이 신설되어 2022년 1월 유동인구 Data가 없는 경우 제외하고 계산하세요.
        - '신림선'의 역들은 2022년 5월에 운영 시작하였습니다.
<br></br>
    - 조건3: 위 1-5를 참고하여 '호선명' Column(열) Data가 '7호선 인천'인 경우는 제외하고 구하세요.
        - 앞서 언급하였다시피 Data에 오류가 있는 것으로 추정되어 제외합니다.
<br></br>
    - 조건4: Index와 Column은 아래와 같이 지정해 주세요.
        - Index: '지하철역'
        - Column: ['유동인구증감률']
<br></br>
    - 조건5: '유동인구증가율' Column(열)을 기준으로 내림차순 정렬하세요.
<br></br>
    - 조건6: 'df_result' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_result)를 check_03_01 함수에 입력하여 채점하세요.

In [290]:
index_num = df_target[df_target['호선명'] == '7호선 인천'].index

In [339]:
df_3_1 = df_target.copy()
df_3_1 =df_3_1.drop(index=index_num)
df_3_1

사용월  호선명  지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자
index                                                               
0      202206  1호선   제기동   218782  263173   209109  281985  20220703
1      202206  1호선   청량리   328288  246164   325365  249818  20220703
2      202206  1호선    종각   834173  131987   815704  121717  20220703
3      202206  1호선   서울역  1150754  194717  1123193  187155  20220703
4      202206  1호선  종로3가   450317  299617   404106  278151  20220703
...       ...  ...   ...      ...     ...      ...     ...       ...
60055  201501  9호선   신방화   169119  160054    38500   38259  20150206
60056  201501  9호선  양천향교   169461  165605    35179   35196  20150206
60057  201501  9호선    등촌   265941  271132    73910   74472  20150206
60058  201501  9호선    가양   469127  439698   107695  104082  20150206
60059  201501  9호선    당산   422698  384841    54179   49891  20150206

[50293 rows x 8 columns]

In [340]:
df_3_1_01 = df_3_1[(df_3_1["사용월"] == 202201)].groupby("지하철역").sum()
df_3_1_01['유동인구'] = df_3_1_01["유임승차인원"] + df_3_1_01["무임승차인원"] + df_3_1_01["유임하차인원"] + df_3_1_01["무임하차인원"]
df_3_1_01

C:\Users\xswer\AppData\Local\Temp\ipykernel_21632\928797882.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df_3_1_01 = df_3_1[(df_3_1["사용월"] == 202201)].groupby("지하철역").sum()


사용월   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자     유동인구
지하철역                                                                 
4.19민주묘지  202201    44418   28668    41966   26036  20220203   141088
가능        202201   119651   45118   115181   44602  20220203   324552
가락시장      404402   309807  118123   325453  118313  40440406   871696
가산디지털단지   404402  1374027  104589  1429932  103901  40440406  3012449
가양        202201   405535  106479   395581  103372  20220203  1010967
...          ...      ...     ...      ...     ...       ...      ...
회기        202201   505073  102760   486758   99849  20220203  1194440
회룡        202201   200765  115995   198983  115075  20220203   630818
회현        202201   449039  135881   481828  136535  20220203  1203283
효창공원앞     404402   200149   42799   195989   42333  40440406   481270
흑석        202201   155338   44906   162385   45541  20220203   408170

[513 rows x 7 columns]

In [341]:
df_3_1_06 = df_3_1[(df_3_1["사용월"] == 202206) & (df_3_1["호선명"] != '신림선')].groupby("지하철역").sum()
df_3_1_06['유동인구'] = df_3_1_06["유임승차인원"] + df_3_1_06["무임승차인원"] + df_3_1_06["유임하차인원"] + df_3_1_06["무임하차인원"]
df_3_1_06

C:\Users\xswer\AppData\Local\Temp\ipykernel_21632\2546074238.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df_3_1_06 = df_3_1[(df_3_1["사용월"] == 202206) & (df_3_1["호선명"] != '신림선')].groupby("지하철역").sum()


사용월   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자     유동인구
지하철역                                                                 
4.19민주묘지  202206    56275   34889    54138   30800  20220703   176102
가능        202206   139026   55693   131155   54568  20220703   380442
가락시장      404412   322711  129590   342636  128815  40441406   923752
가산디지털단지   404412  1431139  117639  1502220  117128  40441406  3168126
가양        202206   456785  123819   442530  120665  20220703  1143799
...          ...      ...     ...      ...     ...       ...      ...
회기        202206   597740  121604   572831  119310  20220703  1411485
회룡        202206   222825  142728   219069  141480  20220703   726102
회현        202206   490517  191008   523086  191242  20220703  1395853
효창공원앞     404412   231459   50062   222409   48748  40441406   552678
흑석        202206   186994   53158   191190   53327  20220703   484669

[513 rows x 7 columns]

In [342]:
df_3_1_06['유동인구증감률'] = round(((df_3_1_06['유동인구'] / df_3_1_01['유동인구']) - 1) * 100, 2)
df_3_1_06

사용월   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자     유동인구  유동인구증감률
지하철역                                                                          
4.19민주묘지  202206    56275   34889    54138   30800  20220703   176102    24.82
가능        202206   139026   55693   131155   54568  20220703   380442    17.22
가락시장      404412   322711  129590   342636  128815  40441406   923752     5.97
가산디지털단지   404412  1431139  117639  1502220  117128  40441406  3168126     5.17
가양        202206   456785  123819   442530  120665  20220703  1143799    13.14
...          ...      ...     ...      ...     ...       ...      ...      ...
회기        202206   597740  121604   572831  119310  20220703  1411485    18.17
회룡        202206   222825  142728   219069  141480  20220703   726102    15.10
회현        202206   490517  191008   523086  191242  20220703  1395853    16.00
효창공원앞     404412   231459   50062   222409   48748  40441406   552678    14.84
흑석        202206   186994   53158   191190   53327  20220703   484669    18.74

[513 rows x 8 columns]

In [343]:
df_3_1_06 = df_3_1_06[df_3_1_06["유동인구증감률"] < 0].drop(columns=["사용월","유임승차인원","무임승차인원","유임하차인원","무임하차인원","작업일자","유동인구"])
df_3_1_06 = df_3_1_06.sort_values(by='유동인구증감률', ascending=False)
df_3_1_06

유동인구증감률
지하철역         
논현      -1.29
신사      -2.74
구반포     -2.87
보라매     -5.36
사릉     -10.15
당고개    -22.42
백양리    -62.19

In [362]:
df_target[(df_target['지하철역']=='보라매') & (df_target['사용월']==202206)]

사용월  호선명 지하철역  유임승차인원  무임승차인원  유임하차인원  무임하차인원      작업일자
index                                                            
230    202206  7호선  보라매  191173   36055  189775   36739  20220703
547    202206  신림선  보라매   53622   17654   49400   15021  20220703

In [352]:
df_3_1_06.drop(index='보라매')

유동인구증감률
지하철역         
논현      -1.29
신사      -2.74
구반포     -2.87
사릉     -10.15
당고개    -22.42
백양리    -62.19

In [355]:
len(df_3_1_06.drop(index='보라매')) == 6,\
sum(df_3_1_06.drop(index='보라매').유동인구증감률 > -5) == 3,\
df_3_1_06.drop(index='보라매').유동인구증감률.sum() < -100,\
df_3_1_06.drop(index='보라매').index[2] == '구반포',\
all([(df_3_1_06.drop(index='보라매').iloc[i, :] > df_3_1_06.drop(index='보라매').iloc[i+1, :]).유동인구증감률 for i in range(len(df_3_1_06.drop(index='보라매'))-1)]),

(True, True, True, True, True)

In [349]:
len(df_3_1_06) == 6,\
sum(df_3_1_06.유동인구증감률 > -5) == 3,\
df_3_1_06.유동인구증감률.sum() < -100,\
df_3_1_06.index[2] == '구반포',\
all([(df_3_1_06.iloc[i, :] > df_3_1_06.iloc[i+1, :]).유동인구증감률 for i in range(len(df_3_1_06)-1)]),

(False, True, True, True, True)

In [356]:
check_03_01(df_3_1_06.drop(index='보라매'))

정답입니다! 15점 누적 되었습니다!
현재 누적 점수: 85 / 100


### 문제 3-2) 심화 정보 얻기 02 (15점)
- 2015년부터 2022년까지 지하철 1~9호선에서 신설된 역을 확인하고자 합니다.
<br></br>
- 위 1단계의 DataFrame(df_target)을 이용하여 아래 예시 및 조건에 맞는 정보를 구하세요.
<br></br>
    - 예시: df_result.iloc[:2]
<br></br>
        <img src="./[문제보기용 이미지]images/08-3-2-001.png" style="margin:auto" width=200/>
<br></br>
    - 조건1: 2015년부터 2022년까지 지하철 1~9호선에서 신설된 역을 확인하고자 합니다.
<br></br>
        - 위 1-5를 참고하여 '호선명' Column(열) Data가 '7호선 인천'인 경우는 제외하고 구하세요.
<br></br>
        - 신설 역: 2015년 1월부터 2022년 6월까지의 기간 중 유동인구('유임승차인원' + '무임승차인원' + '유임하차인원' + '무임하차인원') Data가 존재하지 않는 경우 해당 월에 해당 역은 존재하지 않는다고 판단합니다. 예를들어 2018년 3월부터 유동인구 Data가 있다면 해당 역은 2018년 3월에 신설되었다고 판단합니다.
<br></br>
        - 예시: 9호선 '둔촌오륜'역의 경우 Data가 2018년 12월 부터 2022년 6월까지 존재합니다. 따라서 9호선 '둔촌오륜'역은 2018년 12월에 신설된 역으로 판단합니다.
<br></br>
    - 조건2: Index와 Column은 아래와 같이 지정해 주세요.
        - Index: 0부터 시작하는 연속의 숫자
        - Column: ['호선명', '지하철역', '신설월']
<br></br>
    - 조건3: 아래와 같이 DataFrame을 정렬하고 Index를 초기화 하세요..
        - '호선명' Column(열)을 기준으로 오름차순 정렬하세요.
        - '호선명'이 같다면, '신설월' Column(열)을 기준으로 내림차순 정렬하세요.
        - '호선명'과 '신설월'이 같다면, '지하철역' Column(열)을 기준으로 오름차순 정렬하세요.
        - 위와 같이 정렬 후 Index를 Reset하고, 기존 Index는 삭제(drop)하세요.
<br></br>
    - 조건4: 'df_result' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_result)를 check_03_02 함수에 입력하여 채점하세요.

In [363]:
index_num = df_target[df_target['호선명'] == '7호선 인천'].index

df_3_2 = df_target.copy()
df_3_2 =df_3_2.drop(index=index_num)
df_3_2

사용월  호선명  지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자
index                                                               
0      202206  1호선   제기동   218782  263173   209109  281985  20220703
1      202206  1호선   청량리   328288  246164   325365  249818  20220703
2      202206  1호선    종각   834173  131987   815704  121717  20220703
3      202206  1호선   서울역  1150754  194717  1123193  187155  20220703
4      202206  1호선  종로3가   450317  299617   404106  278151  20220703
...       ...  ...   ...      ...     ...      ...     ...       ...
60055  201501  9호선   신방화   169119  160054    38500   38259  20150206
60056  201501  9호선  양천향교   169461  165605    35179   35196  20150206
60057  201501  9호선    등촌   265941  271132    73910   74472  20150206
60058  201501  9호선    가양   469127  439698   107695  104082  20150206
60059  201501  9호선    당산   422698  384841    54179   49891  20150206

[50293 rows x 8 columns]

In [404]:
df_3_2 = df_3_2[(df_3_2["호선명"]=='1호선') |\
      (df_3_2["호선명"]=='2호선') |\
      (df_3_2["호선명"]=='3호선') |\
      (df_3_2["호선명"]=='4호선') |\
      (df_3_2["호선명"]=='5호선') |\
      (df_3_2["호선명"]=='6호선') |\
      (df_3_2["호선명"]=='7호선') |\
      (df_3_2["호선명"]=='8호선') |\
      (df_3_2["호선명"]=='9호선')]
df_3_2

사용월  호선명  지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자     유동인구
index                                                                        
0      202206  1호선   제기동   218782  263173   209109  281985  20220703   973049
1      202206  1호선   청량리   328288  246164   325365  249818  20220703  1149635
2      202206  1호선    종각   834173  131987   815704  121717  20220703  1903581
3      202206  1호선   서울역  1150754  194717  1123193  187155  20220703  2655819
4      202206  1호선  종로3가   450317  299617   404106  278151  20220703  1432191
...       ...  ...   ...      ...     ...      ...     ...       ...      ...
60055  201501  9호선   신방화   169119  160054    38500   38259  20150206   405932
60056  201501  9호선  양천향교   169461  165605    35179   35196  20150206   405441
60057  201501  9호선    등촌   265941  271132    73910   74472  20150206   685455
60058  201501  9호선    가양   469127  439698   107695  104082  20150206  1120602
60059  201501  9호선    당산   422698  384841    54179   49891  20150206   911609

[27286 rows x 9 columns]

In [405]:
df_3_2['유동인구'] = df_3_2['유임승차인원'] + df_3_2['무임승차인원'] + df_3_2['유임하차인원'] + df_3_2['무임하차인원']
df_3_2

C:\Users\xswer\AppData\Local\Temp\ipykernel_21632\1759633905.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_3_2['유동인구'] = df_3_2['유임승차인원'] + df_3_2['무임승차인원'] + df_3_2['유임하차인원'] + df_3_2['무임하차인원']


사용월  호선명  지하철역   유임승차인원  무임승차인원   유임하차인원  무임하차인원      작업일자     유동인구
index                                                                        
0      202206  1호선   제기동   218782  263173   209109  281985  20220703   973049
1      202206  1호선   청량리   328288  246164   325365  249818  20220703  1149635
2      202206  1호선    종각   834173  131987   815704  121717  20220703  1903581
3      202206  1호선   서울역  1150754  194717  1123193  187155  20220703  2655819
4      202206  1호선  종로3가   450317  299617   404106  278151  20220703  1432191
...       ...  ...   ...      ...     ...      ...     ...       ...      ...
60055  201501  9호선   신방화   169119  160054    38500   38259  20150206   405932
60056  201501  9호선  양천향교   169461  165605    35179   35196  20150206   405441
60057  201501  9호선    등촌   265941  271132    73910   74472  20150206   685455
60058  201501  9호선    가양   469127  439698   107695  104082  20150206  1120602
60059  201501  9호선    당산   422698  384841    54179   49891  20150206   911609

[27286 rows x 9 columns]

In [406]:
df_3_2[df_3_2['지하철역']=="서울역"]['호선명'].unique()

array(['1호선', '4호선'], dtype=object)

In [407]:
list(df_3_2[(df_3_2['지하철역']=="서울역") & (df_3_2['호선명']=="1호선")]['사용월'])[-1]

201501

In [408]:
list(df_3_2[df_3_2['지하철역']=="서울역"]['사용월'])[-1]

201501

In [410]:
line_list = []
station_list = []
start_date = []

station_name = df_3_2['지하철역'].unique()

for name in station_name:
    line_name = df_3_2[df_3_2['지하철역']==name]['호선명'].unique()
    
    for name2 in line_name:
        line_list.append(name2)
        station_list.append(name)
        start_date.append(list(df_3_2[(df_3_2['지하철역']==name) & (df_3_2['호선명']==name2)]['사용월'])[-1])

In [411]:
len(line_list), len(station_list), len(start_date)

(313, 313, 313)

In [445]:
dict = {
    "호선명" : line_list,
    "지하철역" : station_list,
    '신설월' : start_date}

df_3_2 = pd.DataFrame.from_dict(dict)
df_3_2

호선명  지하철역     신설월
0    1호선   제기동  201501
1    1호선   청량리  201501
2    1호선    종각  201501
3    1호선   서울역  201501
4    4호선   서울역  201501
..   ...   ...     ...
308  9호선  둔촌오륜  201812
309  9호선   봉은사  201503
310  9호선  삼성중앙  201503
311  9호선  석촌고분  201812
312  9호선  한성백제  201812

[313 rows x 3 columns]

In [446]:
df_3_2 = df_3_2[df_3_2["신설월"]>201501]
df_3_2 = df_3_2.reset_index().drop(columns='index')
df_3_2

호선명    지하철역     신설월
0   9호선   종합운동장  201503
1   9호선   올림픽공원  201812
2   5호선      강일  202103
3   5호선      미사  202008
4   5호선   하남검단산  202103
5   5호선    하남풍산  202008
6   5호선    하남시청  202103
7   6호선      신내  201912
8   9호선      석촌  201812
9   8호선     남위례  202112
10  9호선      언주  201503
11  9호선  중앙보훈병원  201812
12  9호선    송파나루  201812
13  9호선      삼전  201812
14  9호선     선정릉  201503
15  9호선    둔촌오륜  201812
16  9호선     봉은사  201503
17  9호선    삼성중앙  201503
18  9호선    석촌고분  201812
19  9호선    한성백제  201812

In [447]:
df_3_2 = df_3_2.sort_values(by=["호선명", "신설월", "지하철역"], ascending=[True, False, True])
df_3_2

호선명    지하철역     신설월
2   5호선      강일  202103
4   5호선   하남검단산  202103
6   5호선    하남시청  202103
3   5호선      미사  202008
5   5호선    하남풍산  202008
7   6호선      신내  201912
9   8호선     남위례  202112
15  9호선    둔촌오륜  201812
13  9호선      삼전  201812
8   9호선      석촌  201812
18  9호선    석촌고분  201812
12  9호선    송파나루  201812
1   9호선   올림픽공원  201812
11  9호선  중앙보훈병원  201812
19  9호선    한성백제  201812
16  9호선     봉은사  201503
17  9호선    삼성중앙  201503
14  9호선     선정릉  201503
10  9호선      언주  201503
0   9호선   종합운동장  201503

In [448]:
check_03_02(df_3_4)

오답입니다. 다시 한번 확인해주세요.
현재 누적 점수: 85 / 100


In [450]:
df_target[df_target['지하철역']=='종합운동장']

사용월  호선명   지하철역  유임승차인원  무임승차인원  유임하차인원  무임하차인원      작업일자
index                                                              
53     202206  2호선  종합운동장  342239   51889  380148   53791  20220703
305    202206  9호선  종합운동장  108132   18166  109097   16879  20220703
719    202205  2호선  종합운동장  369686   53307  420587   55198  20220613
1244   202205  9호선  종합운동장  115992   19212  115620   17742  20220613
1890   202204  2호선  종합운동장  264876   42622  296181   44140  20220503
...       ...  ...    ...     ...     ...     ...     ...       ...
59137  201504  9호선  종합운동장   95421   73907    9107    7318  20150508
59160  201503  2호선  종합운동장  420974  450407   57514   58640  20150408
59426  201503  9호선  종합운동장   10537    7589    1710    1572  20150408
59482  201502  2호선  종합운동장  275925  288201   39103   40568  20150311
59784  201501  2호선  종합운동장  336914  351289   47450   48785  20150206

[178 rows x 8 columns]

In [458]:
df_3_2['지하철역'].tolist()

['강일',
 '하남검단산',
 '하남시청',
 '미사',
 '하남풍산',
 '신내',
 '남위례',
 '둔촌오륜',
 '삼전',
 '석촌',
 '석촌고분',
 '송파나루',
 '올림픽공원',
 '중앙보훈병원',
 '한성백제',
 '봉은사',
 '삼성중앙',
 '선정릉',
 '언주',
 '종합운동장']

In [459]:
station_name = df_target[df_target['사용월']==201501]['지하철역'].unique()

for name in station_name:
    for name2 in df_3_2['지하철역'].tolist():
        if name == name2:
            print(name2)

종합운동장
올림픽공원
석촌


In [460]:
df_3_2

호선명    지하철역     신설월
2   5호선      강일  202103
4   5호선   하남검단산  202103
6   5호선    하남시청  202103
3   5호선      미사  202008
5   5호선    하남풍산  202008
7   6호선      신내  201912
9   8호선     남위례  202112
15  9호선    둔촌오륜  201812
13  9호선      삼전  201812
8   9호선      석촌  201812
18  9호선    석촌고분  201812
12  9호선    송파나루  201812
1   9호선   올림픽공원  201812
11  9호선  중앙보훈병원  201812
19  9호선    한성백제  201812
16  9호선     봉은사  201503
17  9호선    삼성중앙  201503
14  9호선     선정릉  201503
10  9호선      언주  201503
0   9호선   종합운동장  201503

In [462]:
df_3_2 = df_3_2.drop(index=[0, 1, 8])
df_3_2 = df_3_2.reset_index(drop=True)
df_3_2 = df_3_2.sort_values(by=["호선명", "신설월", "지하철역"], ascending=[True, False, True])
df_3_2

호선명    지하철역     신설월
0   5호선      강일  202103
1   5호선   하남검단산  202103
2   5호선    하남시청  202103
3   5호선      미사  202008
4   5호선    하남풍산  202008
5   6호선      신내  201912
6   8호선     남위례  202112
7   9호선    둔촌오륜  201812
8   9호선      삼전  201812
9   9호선    석촌고분  201812
10  9호선    송파나루  201812
11  9호선  중앙보훈병원  201812
12  9호선    한성백제  201812
13  9호선     봉은사  201503
14  9호선    삼성중앙  201503
15  9호선     선정릉  201503
16  9호선      언주  201503

In [463]:
check_03_02(df_3_2)

정답입니다! 15점 누적 되었습니다!
현재 누적 점수: 100 / 100
